contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,195 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::fmt::{self, Display, Formatter};
+use std::marker::{Send, Sized};
+
+use rocket::config::{self, Value};
+
+pub use rocket_contrib_codegen::database;
+
+use self::r2d2::ManageConnection;
+
+#[cfg(any(feature = "diesel_sqlite_pool", feature = "diesel_postgres_pool", feature = "diesel_mysql_pool"))]
+pub extern crate diesel;
+
+#[cfg(feature = "postgres_pool")]
+pub extern crate postgres;
+#[cfg(feature = "postgres_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: i64,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T: Display> {
```

Comment:
By stipulating that `T` must implement `Display`, I should be able to get away with printing out this type, I think.

Reply:
This is too onerous of a constraint on `T`. Where is this fact used, anyway? If you have a `DbError<T>` with a concrete `T`, you can already use the fact that `T` is `Display` anywhere you'd like.